src/solvers/thermal_operators.cpp

```diff
@@ -0,0 +1,134 @@
+  m_M_prec   = std::make_unique<mfem::HypreSmoother>();
+
+  m_M_solver->iterative_mode = false;
+  m_M_solver->SetRelTol(params.rel_tol);
+  m_M_solver->SetAbsTol(params.abs_tol);
+  m_M_solver->SetMaxIter(params.max_iter);
+  m_M_solver->SetPrintLevel(params.print_level);
+  m_M_prec->SetType(mfem::HypreSmoother::Jacobi);
+  m_M_solver->SetPreconditioner(*m_M_prec);
+
+  // Use the same options for the T (= M + dt K) solver
+  m_T_solver = std::make_unique<mfem::CGSolver>(m_fespace->GetComm());
+  m_T_prec   = std::make_unique<mfem::HypreSmoother>();
+
+  m_T_solver->iterative_mode = false;
+  m_T_solver->SetRelTol(params.rel_tol);
+  m_T_solver->SetAbsTol(params.abs_tol);
+  m_T_solver->SetMaxIter(params.max_iter);
+  m_T_solver->SetPrintLevel(params.print_level);
+  m_T_solver->SetPreconditioner(*m_T_prec);
+
+  m_state_gf = std::make_shared<mfem::ParGridFunction>(m_fespace.get());
+  m_bc_rhs   = std::make_shared<mfem::Vector>(fespace->GetTrueVSize());
+}
+
+void DynamicConductionOperator::SetMMatrix(std::shared_ptr<mfem::HypreParMatrix> M_mat,
+                                           std::shared_ptr<mfem::HypreParMatrix> M_e_mat)
+{
+  // Set the mass matrix
+  m_M_mat   = M_mat;
+  m_M_e_mat = M_e_mat;
+}
+
+void DynamicConductionOperator::SetKMatrix(std::shared_ptr<mfem::HypreParMatrix> K_mat,
+                                           std::shared_ptr<mfem::HypreParMatrix> K_e_mat)
+{
+  // Set the stiffness matrix and RHS
+  m_K_mat   = K_mat;
+  m_K_e_mat = K_e_mat;
+}
+
+void DynamicConductionOperator::SetLoadVector(std::shared_ptr<mfem::Vector> rhs) { m_rhs = rhs; }
+
+void DynamicConductionOperator::SetEssentialBCs(std::shared_ptr<mfem::Coefficient> ess_bdr_coef,
+                                                mfem::Array<int> &ess_bdr, mfem::Array<int> &ess_tdof_list)
+{
+  m_ess_bdr_coef  = ess_bdr_coef;
+  m_ess_bdr       = ess_bdr;
+  m_ess_tdof_list = ess_tdof_list;
+}
+
+// TODO: allow for changing thermal essential boundary conditions
+void DynamicConductionOperator::Mult(const mfem::Vector &u, mfem::Vector &du_dt) const
+{
+  MFEM_ASSERT(m_M_mat != nullptr, "Mass matrix not set in ConductionSolver::Mult!");
+  MFEM_ASSERT(m_K_mat != nullptr, "Stiffness matrix not set in ConductionSolver::Mult!");
+
+  m_y = u;
+  m_M_solver->SetOperator(*m_M_mat);
+
+  *m_bc_rhs = *m_rhs;
+  mfem::EliminateBC(*m_K_mat, *m_K_e_mat, m_ess_tdof_list, m_y, *m_bc_rhs);
+
+  // Compute:
+  //    du_dt = M^{-1}*-K(u)
+  // for du_dt
+  m_K_mat->Mult(m_y, m_z);
+  m_z.Neg();  // z = -zw  m_z.Add(1.0, *m_bc_rhs);
+  m_z.Add(1.0, *m_bc_rhs);
+  m_M_solver->Mult(m_z, du_dt);
+}
+
+void DynamicConductionOperator::ImplicitSolve(const double dt, const mfem::Vector &u, mfem::Vector &du_dt)
+{
+  MFEM_ASSERT(m_M_mat != nullptr, "Mass matrix not set in ConductionSolver::ImplicitSolve!");
+  MFEM_ASSERT(m_K_mat != nullptr, "Stiffness matrix not set in ConductionSolver::ImplicitSolve!");
+
+  m_y = u;
```

Comment:
That's weird naming to me.... u to y? having no idea what either of those are supposed to be.

Reply:
Fine, I added a comment. 💬

Reply:
Don't get sassy with me...

Reply:
😎